libraries/standard/mqtt/integration-test/mqtt_system_test.c

```diff
@@ -1028,25 +1117,198 @@ void test_MQTT_Restore_Session_Complete_Incoming_Publish( void )
+}
+
+/**
+ * @brief Verifies the behavior of the MQTT library on receiving a duplicate
+ * QoS 1 PUBLISH packet from the broker in a restored session connection.
+ * Tests that the library responds with a PUBACK to for the duplicate incoming QoS 1 PUBLISH
+ * packet that was un-acknowledged in a previous connection of the same session.
+ */
+void test_MQTT_MQTT_Restore_Session_Duplicate_Incoming_Publish_Qos1( void )
+{
+    /* Start a persistent session with the broker. */
+    startPersistentSession();
+
+    /* Subscribe to a topic from which we will be receiving an incomplete incoming
+     * QoS 2 PUBLISH transaction in this connection. */
+    TEST_ASSERT_EQUAL( MQTTSuccess, subscribeToTopic(
+                           &context, TEST_MQTT_TOPIC, MQTTQoS1 ) );
+    TEST_ASSERT_FALSE( receivedSubAck );
+    TEST_ASSERT_EQUAL( MQTTSuccess,
+                       MQTT_ProcessLoop( &context, MQTT_PROCESS_LOOP_TIMEOUT_MS ) );
+    TEST_ASSERT_TRUE( receivedSubAck );
+
+    /* Publish to the same topic with Qos 1 (so that the broker can re-publish it back to us). */
+    TEST_ASSERT_EQUAL( MQTTSuccess, publishToTopic(
+                           &context, TEST_MQTT_TOPIC, false, MQTTQoS1, MQTT_GetPacketId( &context ) ) );
+
+    /* Disconnect on receiving the incoming PUBLISH packet from the broker so that
+     * an acknowledgement cannot be sent to the broker. */
+    disconnectOnPacketType = MQTT_PACKET_TYPE_PUBLISH;
+    TEST_ASSERT_EQUAL( MQTTSendFailed,
+                       MQTT_ProcessLoop( &context, 2 * MQTT_PROCESS_LOOP_TIMEOUT_MS ) );
+
+    /* Make sure that a record was created for the incoming PUBLISH packet. */
+    TEST_ASSERT_NOT_EQUAL( MQTT_PACKET_ID_INVALID, context.incomingPublishRecords[ 0 ].packetId );
+
+    /* Verify that the connection with the broker has been disconnected. */
+    TEST_ASSERT_EQUAL( MQTTNotConnected, context.connectStatus );
+
+    /* We will re-establish an MQTT over TLS connection with the broker to restore
+     * the persistent session. */
+    resumePersistentSession();
+
+    /* Clear the global variable for not disconnecting on the duplicate PUBLISH
+     * packet that we receive from the broker on the session restoration. */
     disconnectOnPacketType = MQTT_PACKET_TYPE_INVALID;
 
-    /* Resume the incomplete incoming QoS 2 PUBLISH transaction from the previous MQTT connection. */
+    /* Process the duplicate incoming QoS 1 PUBLISH that will be sent by the broker
+     * to re-attempt the PUBLISH operation. */
+    TEST_ASSERT_EQUAL( MQTTSuccess,
+                       MQTT_ProcessLoop( &context, 2 * MQTT_PROCESS_LOOP_TIMEOUT_MS ) );
+
+    /* Make sure that the library cleared the record for the incoming QoS 1 PUBLISH packet. */
+    TEST_ASSERT_EQUAL( MQTT_PACKET_ID_INVALID, context.incomingPublishRecords[ 0 ].packetId );
+}
+
+/**
+ * @brief Verifies the behavior of the MQTT library on receiving a duplicate
+ * QoS2 PUBLISH packet from the broker in a restored session connection.
```

Comment:
Looks a little weird that `QoS2` is written here but `QoS 2` elsewhere.
```suggestion
 * QoS 2 PUBLISH packet from the broker in a restored session connection.
```